### PR TITLE
New protocol OPC-DA and update for modbus

### PR DESCRIPTION
**Pull request by G4RUF**
I create a file to analyze opc-da, it is not completed but I still working on it. It is usefull to have a first experience in opc-da.
I add some modifications in modbus to complete my experience on it and my colleague.

### CODE REVIEW

File: scapy/contrib/modbus.py [hunk 1]
```diff
@@ -14,11 +14,9 @@
 # You should have received a copy of the GNU General Public License
 # along with Scapy. If not, see <http://www.gnu.org/licenses/>.
 
-# Copyright (C) 2016 Arthur Gervais, Ken LE PRADO, Sébastien Mainand
+# Copyright (C) 2016 Arthur Gervais, Ken LE PRADO, Sébastien Mainand
 
-from scapy.packet import *
-from scapy.fields import *
-from scapy.layers.inet import *
+from scapy.all import *
```
Comment (gpotter2) [inline scapy/contrib/modbus.py]: Is that really useful ? Wouldn't it be better to only import what you need ?

File: .gitignore [hunk 2]
```diff
@@ -5,3 +5,4 @@ build/
 MANIFEST
 *.egg-info/
 scapy/VERSION
+scapy/contrib/opc_da_ontest.py
```
Comment (gpotter2) [inline .gitignore]: Are you sure this is wanted ?

Comment (G4RUF) [inline .gitignore]: it was just for me when I tested opc da and I didn't want pollute the code with my tests.

Comment (p-l-) [inline .gitignore]: So maybe you shouldn't pollute the repository's .gitignore for your tests?

File: scapy/contrib/modbus.uts [hunk 3]
```diff
@@ -13,34 +13,43 @@ str(ModbusADURequest() / '\x00\x01\x02') == '\x00\x00\x00\x00\x00\x04\xff\x00\x0
 
 = MBAP Guess Payload ModbusPDU01ReadCoilsRequest (simple case)
 p = ModbusADURequest('\x00\x00\x00\x00\x00\x06\xff\x01\x00\x00\x00\x01')
-assert(isinstance(p.payload, ModbusPDU01ReadCoilsRequest))
+p.payload.__class__.__name__ == 'ModbusPDU01ReadCoilsRequest'
```
Comment (p-l-) [inline scapy/contrib/modbus.uts]: Why would you do this?

Comment (p-l-) [inline scapy/contrib/modbus.uts]: Why would you do this?

File: .gitignore (hunk 2 above)
Comment (G4RUF) [inline .gitignore]: Ok. Sorry, should I push with a non polluting .gitignore ?

File: scapy/contrib/modbus.py (hunk 1 above)
Comment (G4RUF) [inline scapy/contrib/modbus.py]: I do my test again and no it's not useful. The old imports are only what we need. I missed this.

File: scapy/contrib/modbus.uts (hunk 3 above)
Comment (G4RUF) [inline scapy/contrib/modbus.uts]: A bad merge from an old scapy and mine. I didn't notice it was this kind of modifications.

**Comment by gpotter2** [comment]
@G4RUF Hello. If you still want this PR to be taken in consideration, please update your code to match the current scapy version's requirements, otherwise this will be closed for inactivity.

**Comment by G4RUF** [comment]
@gpotter2 Hello. Is it okay with this PR or do I make some corrections ?

**Comment by gpotter2** [comment]
It looks relevent.
What do you think @p-l- ?

File: scapy/contrib/modbus.py [hunk 4]
```diff
@@ -571,6 +535,162 @@ class ModbusPDU2B0EReadDeviceIdentificationError(Packet):
                    ByteEnumField("exceptCode", 1, _modbus_exceptions)]
 
 
+_reserved_funccode_request = {
+    0x08: '0x08 Unknown Reserved Request',
```
Comment (gpotter2) [inline scapy/contrib/modbus.py]: I don't really like leaving some work to do like this... Can't you find the fields ?

File: scapy/contrib/modbus.py [hunk 5]
```diff
@@ -571,6 +535,162 @@ class ModbusPDU2B0EReadDeviceIdentificationError(Packet):
...
+    0x8E: '0x8e Unknown Reserved Error',
+    0xA9: '0x88 Unknown Reserved Error',
+    0xAA: '0x88 Unknown Reserved Error',
+    0xDA: 'Specific Schneider Electric Error',
+    0xDB: '0xdb Unknown Reserved Error',
+    0xDC: '0xdc Unknown Reserved Error',
+    0xFD: '0xfd Unknown Reserved Error',
+    0xFE: '0xfe Unknown Reserved Error',
+    0xFF: '0xff Unknown Reserved Error',
+        }
```
Comment (gpotter2) [inline scapy/contrib/modbus.py]: Can you fix the indentation ?

**Comment by smainand** [comment]
Hello,  
I checked with @G4RUF and the Modbus part of this PR has already been merged to Scapy.
The OPC-DA part is now included in the PR #1384, so this one can be safely closed.
Seb.

**Comment by guedou** [comment]
Thanks for your help!

**Pull request closed without merging**